### ISSUE TITLE
append_if_no_such_line recreates file with wrong file ower

### ISSUE THREAD

**Issue by bollwarm** [RexOps/Rex#1455]
## Describe the bug

append_if_no_such_line recreates file with wrong file ower，when using  append_if_no_such_line function  sudo  to add a line for sudoers

## How to reproduce it

user rex  sudo on 
and a Task 
```perl
task "filetest", group =>"all", sub {

append_if_no_such_line "/etc/sudoers", "aa ALL=(ALL) ALL";

};
```

## Expected behavior

[2020-11-10 15:28:52] ERROR - Error executing task:
[2020-11-10 15:28:52] ERROR - malformed JSON string, neither tag, array, object, number, string or atom, at character offset 0 (before "sudo: /etc/sudoers i...") at /usr/local/share/perl5/Rex/Interface/Fs/Sudo.pm line 178.

[2020-11-10 15:28:53] ERROR - 1 out of 1 task(s) failed:
[2020-11-10 15:28:53] ERROR -   useradd failed on xx.xx.xx..162
[2020-11-10 15:28:53] ERROR -           malformed JSON string, neither tag, array, object, number, string or atom, at character offset 0 (before "sudo: /etc/sudoers i...") at /usr/local/share/perl5/Rex/Interface/Fs/Sudo.pm line 178.

And then the /etc/sudoers owner be change to user not root ，and then we can not using sudo。

## Circumstances

 - Rex version:  1.7.0 and 1.13.1
 - Perl version:  5.16
 - OS running rex:  centos 
 - OS managed by rex: 
 - How rex was installed: cpanm

**Comment by ferki** [comment]
## TL;DR

I believe this issue is a mix of at least two topics:

- sudoers handling should take special considerations into account regardless of the workflow used to manage it
- the current overall file handling flow of Rex is highly unlikely to be changed because of this special case

## Sudoers handling

Specifically in the case of managing `/etc/sudoers` file manually, one should use `visudo` to avoid similar situations (locking themselves out of sudo due to corrupted sudoers config content/permissions). What `visudo` does is roughly:

- it locks the file, so only one process may edit sudoers at a time
- takes a copy of `/etc/sudoers` as `/etc/sudoers.tmp`
- opens the temp file for editing
- upon saving, it validates the temporary file
- copies the new content back to `/etc/sudoers` if the contents are valid

As with other common configuration management solutions, the same procedure should be followed when editing `/etc/sudoers` programmatically as well. With Rex that could look like more or less the following:

```perl
task 'update_sudoers', sub {
    my $sudoers     = '/etc/sudoers';
    my $tmp_sudoers = '/etc/sudoers.tmp';

    # copy sudoers
    cp $sudoers, $tmp_sudoers;

    # edit the copy
    append_if_no_such_line $tmp_sudoers, 'aa ALL=(ALL) ALL';

    # apply sudoers only if it's valid
    run "cp $tmp_sudoers $sudoers",
      only_if => "visudo --check --file $tmp_sudoers";

    # clean up sudoers copy
    file $tmp_sudoers,
      ensure => 'absent';
};
```

Please note I only tested the above on a current up-to-date Gentoo system. Also variations of the same technique might be useful too, like:

- actually adding a file lock at the beginning and unlock at the end
- using different temp file path
- preparing a run command with `only_notified => TRUE` to be notified later
- making the decisions as part of an `if..then..else`
- etc.

## File handling in Rex

As far as file operation handling in Rex is concerned, the flow is roughly the following (which is comparable to the flow of other tools, even newer ones than Rex):

- Rex operates on a temp file instead of directly on the target file itself, because it cannot know in advance if the operations will be successful (e.g. on `/etc/.rex.tmp.sudoers` instead of on `/etc/sudoers`)
- when the write operations on the Rex temp file are finished, it takes a checksum of the file, and compares it with the original
- if the checksums are the same, then nothings changes, so discards the temp file, and returns
- if the checksums are different, then the content has changed, so it moves the tempfile over to the target file
- restores file attributes on the target file if needed (owner, group, permissions), and returns

Generally speaking, the setting of file attributes is the last step in this process in order to avoid any issues upon copy/move due to wrong ownership/permission settings. I'd say this flow is highly unlikely to change due to high risk of running into even more corner cases, and breaking backwards compatibility. If higher control is needed over the process, for example in the case of sudoers file, or files containing sensitive information, it's always possible to implement the custom logic inside the task.

I hope the above helps with your use case already, and I'm open to discuss further alternatives.

**Comment by bollwarm** [comment]
Think You,I will try the method.

**Comment by bollwarm** [comment]
The method is OK！ In  Centos ，It‘s  successed。  Thank you very much！
